File: apps/go-api/internal/store/influx.go
package store

import (
	"bytes"
	"context"
	"encoding/csv"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	influxdb2 "github.com/influxdata/influxdb-client-go/v2"

	"prism/go-api/internal/config"
	"prism/go-api/internal/model"
)

type Influx struct {
	client   influxdb2.Client
	http     *http.Client
	base     string
	writeURL string
	token    string
	org      string
	bucket   string
	tags     catalogMem
}

func NewInflux(cfg config.Config) (*Influx, error) {
	base := strings.TrimRight(cfg.InfluxURL, "/")
	client := influxdb2.NewClient(cfg.InfluxURL, cfg.InfluxToken)
	writeURL := base + "/api/v2/write?" + url.Values{
		"org":       {cfg.InfluxOrg},
		"bucket":    {cfg.InfluxBucket},
		"precision": {"ns"},
	}.Encode()
	return &Influx{
		client:   client,
		http:     newWriteHTTPClient(30 * time.Second),
		base:     base,
		writeURL: writeURL,
		token:    cfg.InfluxToken,
		org:      cfg.InfluxOrg,
		bucket:   cfg.InfluxBucket,
	}, nil
}

func (s *Influx) Name() string { return "influxdb" }

func (s *Influx) Close() error {
	s.client.Close()
	s.http.CloseIdleConnections()
	return nil
}

func (s *Influx) Ping(ctx context.Context) error {
	ok, err := s.client.Ping(ctx)
	if err != nil {
		return err
	}
	if !ok {
		return errors.New("influxdb ping failed")
	}
	return s.ensureDBRP(ctx)
}

type influxDBRPList struct {
	Content []json.RawMessage `json:"content"`
}

type influxBucketList struct {
	Buckets []struct {
		ID   string `json:"id"`
		Name string `json:"name"`
	} `json:"buckets"`
}

func (s *Influx) ensureDBRP(ctx context.Context) error {
	q := url.Values{"org": {s.org}, "db": {s.bucket}}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.base+"/api/v2/dbrps?"+q.Encode(), nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Token "+s.token)
	resp, err := s.http.Do(req)
	if err != nil {
		return err
	}
	body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	closeHTTP(resp)
	if resp.StatusCode < 300 {
		var listed influxDBRPList
		if err := json.Unmarshal(body, &listed); err == nil && len(listed.Content) > 0 {
			return nil
		}
	}
	req, err = http.NewRequestWithContext(ctx, http.MethodGet, s.base+"/api/v2/buckets?"+url.Values{"org": {s.org}, "name": {s.bucket}}.Encode(), nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Token "+s.token)
	resp, err = s.http.Do(req)
	if err != nil {
		return err
	}
	body, _ = io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	closeHTTP(resp)
	if resp.StatusCode >= 300 {
		return fmt.Errorf("influx buckets %d: %s", resp.StatusCode, body)
	}
	var buckets influxBucketList
	if err := json.Unmarshal(body, &buckets); err != nil || len(buckets.Buckets) == 0 {
		return fmt.Errorf("influx bucket %q not found", s.bucket)
	}
	payload, _ := json.Marshal(map[string]any{
		"org":              s.org,
		"bucketID":         buckets.Buckets[0].ID,
		"database":         s.bucket,
		"retention_policy": "autogen",
		"default":          true,
	})
	req, err = http.NewRequestWithContext(ctx, http.MethodPost, s.base+"/api/v2/dbrps?org="+url.QueryEscape(s.org), bytes.NewReader(payload))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Token "+s.token)
	req.Header.Set("Content-Type", "application/json")
	resp, err = s.http.Do(req)
	if err != nil {
		return err
	}
	body, _ = io.ReadAll(io.LimitReader(resp.Body, 2048))
	closeHTTP(resp)
	if resp.StatusCode >= 300 && resp.StatusCode != http.StatusConflict {
		return fmt.Errorf("influx dbrp %d: %s", resp.StatusCode, body)
	}
	return nil
}

func (s *Influx) Write(ctx context.Context, samples []model.Sample) error {
	if len(samples) == 0 {
		return nil
	}
	buf := getBuf()
	defer putBuf(buf)
	for i := range samples {
		appendInfluxLine(buf, &samples[i])
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.writeURL, bytes.NewReader(buf.Bytes()))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Token "+s.token)
	req.Header.Set("Content-Type", "text/plain; charset=utf-8")
	resp, err := s.http.Do(req)
	if err != nil {
		return err
	}
	defer closeHTTP(resp)
	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return fmt.Errorf("influx write %d: %s", resp.StatusCode, body)
	}
	return nil
}

func appendInfluxLine(buf *bytes.Buffer, p *model.Sample) {
	buf.WriteString("samples,tag_id=")
	buf.WriteString(strconv.FormatUint(uint64(p.TagID), 10))
	buf.WriteString(" value=")
	buf.WriteString(strconv.FormatFloat(p.Value, 'g', -1, 64))
	buf.WriteString(",quality=")
	buf.WriteString(strconv.FormatUint(uint64(p.Quality), 10))
	buf.WriteString("i ")
	buf.WriteString(strconv.FormatInt(p.TS.UTC().UnixNano(), 10))
	buf.WriteByte('\n')
}

func (s *Influx) Locf(ctx context.Context, tagIDs []uint32, at time.Time) ([]model.Sample, error) {
	return s.queryLast(ctx, tagIDs, at.UTC(), false)
}

func (s *Influx) Range(ctx context.Context, tagIDs []uint32, from, to time.Time) ([]model.Sample, error) {
	seed, err := s.queryLast(ctx, tagIDs, from.UTC(), true)
	if err != nil {
		return nil, err
	}
	mid, err := s.queryWindow(ctx, tagIDs, from.UTC(), to.UTC())
	if err != nil {
		return nil, err
	}
	return append(seed, mid...), nil
}

func (s *Influx) queryLast(ctx context.Context, tagIDs []uint32, stop time.Time, carried bool) ([]model.Sample, error) {
	q := fmt.Sprintf(
		`SELECT last("value") AS "value", last("quality") AS "quality" FROM "samples" WHERE time <= %s AND %s GROUP BY "tag_id"`,
		influxQLTime(stop), influxQLTagRE(tagIDs),
	)
	return s.queryInfluxQL(ctx, q, carried)
}

func (s *Influx) queryWindow(ctx context.Context, tagIDs []uint32, from, to time.Time) ([]model.Sample, error) {
	q := fmt.Sprintf(
		`SELECT "value", "quality" FROM "samples" WHERE time > %s AND time <= %s AND %s`,
		influxQLTime(from), influxQLTime(to), influxQLTagRE(tagIDs),
	)
	return s.queryInfluxQL(ctx, q, false)
}

type influxQLResp struct {
	Results []struct {
		Error  string `json:"error"`
		Series []struct {
			Tags    map[string]string `json:"tags"`
			Columns []string          `json:"columns"`
			Values  [][]any           `json:"values"`
		} `json:"series"`
	} `json:"results"`
}

func (s *Influx) queryInfluxQL(ctx context.Context, q string, carried bool) ([]model.Sample, error) {
	form := url.Values{
		"org":    {s.org},
		"bucket": {s.bucket},
		"db":     {s.bucket},
		"epoch":  {"ms"},
		"q":      {q},
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.base+"/query", strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Token "+s.token)
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/csv")
	resp, err := s.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer closeHTTP(resp)
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("influx query %d: %s", resp.StatusCode, body)
	}
	if len(body) == 0 || body[0] != '{' {
		return parseInfluxCSV(body, carried)
	}
	var parsed influxQLResp
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, err
	}
	out := make([]model.Sample, 0, 64)
	for _, result := range parsed.Results {
		if result.Error != "" {
			return nil, fmt.Errorf("influxql: %s", result.Error)
		}
		for _, series := range result.Series {
			id, _ := strconv.ParseUint(series.Tags["tag_id"], 10, 32)
			col := map[string]int{}
			for i, name := range series.Columns {
				col[name] = i
			}
			ti, okT := col["time"]
			vi, okV := col["value"]
			qi, hasQ := col["quality"]
			if !okT || !okV {
				continue
			}
			for _, row := range series.Values {
				if ti >= len(row) || vi >= len(row) {
					continue
				}
				ts := influxQLTS(row[ti])
				val := asFloat(row[vi])
				q := uint16(0)
				if hasQ && qi < len(row) {
					q = uint16(asFloat(row[qi]))
				}
				out = append(out, model.Sample{TS: ts, TagID: uint32(id), Value: val, Quality: q, Carried: carried})
			}
		}
	}
	return out, nil
}

func influxQLTime(t time.Time) string {
	return "'" + t.UTC().Format(time.RFC3339Nano) + "'"
}

func influxQLTagRE(ids []uint32) string {
	if len(ids) == 0 {
		return "true"
	}
	parts := make([]string, len(ids))
	for i, id := range ids {
		parts[i] = strconv.FormatUint(uint64(id), 10)
	}
	return `tag_id =~ /^(` + strings.Join(parts, "|") + `)$/`
}

func influxQLTS(v any) time.Time {
	switch t := v.(type) {
	case float64:
		return time.UnixMilli(int64(t)).UTC()
	case json.Number:
		n, _ := t.Int64()
		return time.UnixMilli(n).UTC()
	case string:
		if ts, err := time.Parse(time.RFC3339Nano, t); err == nil {
			return ts.UTC()
		}
		if n, err := strconv.ParseInt(t, 10, 64); err == nil {
			return time.UnixMilli(n).UTC()
		}
	}
	return time.Time{}
}

func parseInfluxCSV(body []byte, carried bool) ([]model.Sample, error) {
	cr := csv.NewReader(bytes.NewReader(body))
	cr.ReuseRecord = true
	cr.LazyQuotes = true
	cr.FieldsPerRecord = -1
	idx := map[string]int{}
	header := false
	out := make([]model.Sample, 0, 1024)
	for {
		row, err := cr.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		if len(row) == 0 || strings.HasPrefix(row[0], "#") {
			continue
		}
		if !header {
			for i, name := range row {
				idx[strings.ToLower(strings.TrimSpace(name))] = i
			}
			if _, ok := idx["time"]; !ok {
				continue
			}
			header = true
			continue
		}
		ti, okT := idx["time"]
		vi, okV := idx["value"]
		if !okT || !okV || ti >= len(row) || vi >= len(row) {
			continue
		}
		id := uint32(0)
		if i, ok := idx["tag_id"]; ok && i < len(row) {
			n, _ := strconv.ParseUint(row[i], 10, 32)
			id = uint32(n)
		} else if i, ok := idx["tags"]; ok && i < len(row) {
			id = tagIDFromInfluxTags(row[i])
		}
		q := uint16(0)
		if i, ok := idx["quality"]; ok && i < len(row) {
			f, _ := strconv.ParseFloat(row[i], 64)
			q = uint16(f)
		}
		val, _ := strconv.ParseFloat(row[vi], 64)
		out = append(out, model.Sample{
			TS: influxQLTS(row[ti]), TagID: id, Value: val, Quality: q, Carried: carried,
		})
	}
	return out, nil
}

func tagIDFromInfluxTags(s string) uint32 {
	s = strings.Trim(s, `"`)
	for _, part := range strings.Split(s, ",") {
		k, v, ok := strings.Cut(strings.TrimSpace(part), "=")
		if ok && k == "tag_id" {
			n, _ := strconv.ParseUint(v, 10, 32)
			return uint32(n)
		}
	}
	return 0
}

func (s *Influx) UpsertTags(_ context.Context, tags []model.Tag) error {
	s.tags.upsert(tags)
	return nil
}

func (s *Influx) ListTags(_ context.Context) ([]model.Tag, error) {
	return s.tags.list(), nil
}

type catalogMem struct {
	mu   sync.Mutex
	data map[uint32]model.Tag
}

func (c *catalogMem) upsert(tags []model.Tag) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.data == nil {
		c.data = map[uint32]model.Tag{}
	}
	for _, t := range tags {
		c.data[t.ID] = t
	}
}

func (c *catalogMem) list() []model.Tag {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make([]model.Tag, 0, len(c.data))
	for _, t := range c.data {
		out = append(out, t)
	}
	return out
}

File: apps/python-api/app/store/influx.py
from datetime import datetime, timezone
from urllib.parse import quote
import csv
import io

from app.models import Sample, Tag
from app.store.catalog import CatalogMem
from app.store.net import ilp_float, new_client, raise_status, unix_ns


def _influxql_time(ts: datetime) -> str:
    utc = ts.astimezone(timezone.utc) if ts.tzinfo else ts.replace(tzinfo=timezone.utc)
    return "'" + utc.isoformat().replace("+00:00", "Z") + "'"


def _tag_id_from_tags(raw: str) -> int:
    for part in raw.strip('"').split(","):
        if "=" in part:
            key, val = part.split("=", 1)
            if key.strip() == "tag_id":
                return int(val)
    return 0


def _parse_influx_csv(text: str, carried: bool) -> list[Sample]:
    if not text:
        return []
    reader = csv.reader(io.StringIO(text))
    idx: dict[str, int] | None = None
    out: list[Sample] = []
    for row in reader:
        if not row or row[0].startswith("#"):
            continue
        if idx is None:
            candidate = {name.strip().lower(): i for i, name in enumerate(row)}
            if "time" not in candidate:
                continue
            idx = candidate
            continue
        ti, vi = idx.get("time"), idx.get("value")
        if ti is None or vi is None or ti >= len(row) or vi >= len(row):
            continue
        if "tag_id" in idx and idx["tag_id"] < len(row) and row[idx["tag_id"]]:
            tag_id = int(float(row[idx["tag_id"]]))
        elif "tags" in idx and idx["tags"] < len(row):
            tag_id = _tag_id_from_tags(row[idx["tags"]])
        else:
            tag_id = 0
        quality = 0
        if "quality" in idx and idx["quality"] < len(row) and row[idx["quality"]]:
            quality = int(float(row[idx["quality"]]))
        ts_raw = row[ti]
        try:
            ms = int(float(ts_raw))
            ts = datetime.fromtimestamp(ms / 1000, tz=timezone.utc)
        except ValueError:
            ts = datetime.fromisoformat(ts_raw.replace("Z", "+00:00"))
        out.append(
            Sample.model_construct(
                ts=ts, tag_id=tag_id, value=float(row[vi] or 0), quality=quality, carried=carried
            )
        )
    return out


def _tag_re(tag_ids: list[int]) -> str:
    if not tag_ids:
        return "true"
    return "tag_id =~ /^(" + "|".join(str(i) for i in tag_ids) + ")$/"


class InfluxStore:
    name = "influxdb"

    def __init__(self, url: str, token: str, org: str, bucket: str) -> None:
        self._org = org
        self._bucket = bucket
        self._write_path = (
            f"api/v2/write?org={quote(org)}&bucket={quote(bucket)}&precision=ns"
        )
        self._http = new_client(
            timeout=30.0,
            base_url=url.rstrip("/") + "/",
            headers={"Authorization": f"Token {token}"},
        )
        self._tags = CatalogMem()

    async def ping(self) -> None:
        resp = await self._http.get("health")
        if resp.status_code >= 300:
            raise RuntimeError("influxdb ping failed")
        await self._ensure_dbrp()

    async def write(self, samples: list[Sample]) -> None:
        if not samples:
            return
        parts = [
            f"samples,tag_id={s.tag_id} value={ilp_float(s.value)},quality={s.quality}i {unix_ns(s.ts)}\n"
            for s in samples
        ]
        resp = await self._http.post(
            self._write_path,
            content="".join(parts).encode("ascii"),
            headers={"Content-Type": "text/plain; charset=utf-8"},
        )
        raise_status(resp, "influxdb write")

    async def locf(self, tag_ids: list[int], at: datetime) -> list[Sample]:
        return await self._last(tag_ids, at, False)

    async def range(self, tag_ids: list[int], start: datetime, end: datetime) -> list[Sample]:
        seed = await self._last(tag_ids, start, True)
        q = (
            f'SELECT "value", "quality" FROM "samples" '
            f"WHERE time > {_influxql_time(start)} AND time <= {_influxql_time(end)} AND {_tag_re(tag_ids)}"
        )
        return seed + await self._query(q, False)

    async def _last(self, tag_ids: list[int], stop: datetime, carried: bool) -> list[Sample]:
        q = (
            f'SELECT last("value") AS "value", last("quality") AS "quality" FROM "samples" '
            f"WHERE time <= {_influxql_time(stop)} AND {_tag_re(tag_ids)} "
            f'GROUP BY "tag_id"'
        )
        return await self._query(q, carried)

    async def _query(self, q: str, carried: bool) -> list[Sample]:
        resp = await self._http.post(
            "query",
            data={"org": self._org, "bucket": self._bucket, "db": self._bucket, "epoch": "ms", "q": q},
            headers={"Accept": "application/csv"},
        )
        raise_status(resp, "influxdb query")
        text = resp.text
        if not text or text[:1] != "{":
            return _parse_influx_csv(text, carried)
        payload = resp.json()
        out: list[Sample] = []
        for result in payload.get("results") or []:
            if result.get("error"):
                raise RuntimeError(result["error"])
            for series in result.get("series") or []:
                tag_id = int((series.get("tags") or {}).get("tag_id") or 0)
                cols = {name: i for i, name in enumerate(series.get("columns") or [])}
                ti, vi = cols.get("time"), cols.get("value")
                qi = cols.get("quality")
                if ti is None or vi is None:
                    continue
                for row in series.get("values") or []:
                    if ti >= len(row) or vi >= len(row):
                        continue
                    ts = datetime.fromtimestamp(int(row[ti]) / 1000, tz=timezone.utc)
                    quality = int(float(row[qi])) if qi is not None and qi < len(row) else 0
                    out.append(
                        Sample.model_construct(
                            ts=ts,
                            tag_id=tag_id,
                            value=float(row[vi]),
                            quality=quality,
                            carried=carried,
                        )
                    )
        return out

    async def _ensure_dbrp(self) -> None:
        listed = await self._http.get("api/v2/dbrps", params={"org": self._org, "db": self._bucket})
        if listed.status_code < 300:
            content = (listed.json() or {}).get("content") or []
            if content:
                return
        buckets = await self._http.get("api/v2/buckets", params={"org": self._org, "name": self._bucket})
        raise_status(buckets, "influx buckets")
        items = (buckets.json() or {}).get("buckets") or []
        if not items:
            raise RuntimeError(f"influx bucket {self._bucket} not found")
        resp = await self._http.post(
            "api/v2/dbrps",
            params={"org": self._org},
            json={
                "org": self._org,
                "bucketID": items[0]["id"],
                "database": self._bucket,
                "retention_policy": "autogen",
                "default": True,
            },
        )
        if resp.status_code >= 300 and resp.status_code != 409:
            raise_status(resp, "influx dbrp")

    async def upsert_tags(self, tags: list[Tag]) -> None:
        self._tags.upsert(tags)

    async def list_tags(self) -> list[Tag]:
        return self._tags.list()

    async def close(self) -> None:
        await self._http.aclose()
